sklearn/kernel_approximation.py

```diff
@@ -494,3 +498,200 @@ def _get_kernel_params(self):
+
+    Notes
+    -----
+    See "Fastfood | Approximating Kernel Expansions in Loglinear Time" by
+    Quoc Le, Tamas Sarl and Alex Smola.
+
+    Examples
+    ----
+    See scikit-learn-fastfood/examples/plot_digits_classification_fastfood.py
+    for an example how to use fastfood with a primal classifier in comparison
+    to an usual rbf-kernel with a dual classifier.
+
+    """
+
+    def __init__(self,
+                 sigma=np.sqrt(1/2),
+                 n_components=100,
+                 tradeoff_mem_accuracy='accuracy',
+                 random_state=None):
+        self.sigma = sigma
+        self.n_components = n_components
+        self.random_state = random_state
+        self.rng = check_random_state(self.random_state)
+        # map to 2*n_components features or to n_components features with less
+        # accuracy
+        self.tradeoff_mem_accuracy = \
+            tradeoff_mem_accuracy
+
+    @staticmethod
+    def is_number_power_of_two(n):
+        return n != 0 and ((n & (n - 1)) == 0)
+
+    @staticmethod
+    def enforce_dimensionality_constraints(d, n):
+        if not (Fastfood.is_number_power_of_two(d)):
+            # find d that fulfills 2^l
+            d = np.power(2, np.floor(np.log2(d)) + 1)
+        divisor, remainder = divmod(n, d)
+        times_to_stack_v = int(divisor)
+        if remainder != 0:
+            # output info, that we increase n so that d is a divider of n
+            n = (divisor + 1) * d
+            times_to_stack_v = int(divisor+1)
+        return int(d), int(n), times_to_stack_v
+
+    def pad_with_zeros(self, X):
+        try:
+            X_padded = np.pad(X,
+                              ((0, 0),
+                               (0, self.number_of_features_to_pad_with_zeros)),
+                              'constant')
+        except AttributeError:
+            zeros = np.zeros((X.shape[0],
+                              self.number_of_features_to_pad_with_zeros))
+            X_padded = np.concatenate((X, zeros), axis=1)
+
+        return X_padded
+
+    @staticmethod
+    def approx_fourier_transformation_multi_dim(result):
+        cyfht(result)
+
+    @staticmethod
+    def l2norm_along_axis1(X):
+        return np.sqrt(np.einsum('ij,ij->i', X, X))
+
+    def uniform_vector(self):
+        if self.tradeoff_mem_accuracy != 'accuracy':
+            return self.rng.uniform(0, 2 * np.pi, size=self.n)
+        else:
+            return None
+
+    def apply_approximate_gaussian_matrix(self, B, G, P, X):
+        """ Create mapping of all x_i by applying B, G and P step-wise """
+        num_examples = X.shape[0]
+
+        result = np.multiply(B, X.reshape((1, num_examples, 1, self.d)))
+        result = result.reshape((num_examples*self.times_to_stack_v, self.d))
+        Fastfood.approx_fourier_transformation_multi_dim(result)
+        result = result.reshape((num_examples, -1))
+        np.take(result, P, axis=1, mode='wrap', out=result)
```

Comment:
@dougalsutherland 
Indeed it does not do what I expected. But thinking about it makes me agree its behavior is right even it is unexpected. Changing the permute operation from non-inplace to inplace shouldn't change the outcome.
To summarize the observed behavior: 
```
result = np.arange(0.4)
permutation = [1, 2, 3, 0, 4]
np.take(result, permutation,mode='wrap',out=result) 
```
is expected to be [1, 2, 3, 0, 4] but yields [1, 2, 3, 1, 4] because the element at index 3 of permutation refers to an already changed value in result.

After this I think there is no way around copying the array?!